yarn-project/simulator/src/avm/avm_tree.ts

```diff
@@ -0,0 +1,808 @@
+    // This is needed since we have a sparse collectio of "value" sorted leaves in the ephemeral tree
+    public indexedUpdates: Map<IndexedTreeId, Map<bigint, Buffer>>,
+  ) {}
+
+  static async create(treeDb: MerkleTreeReadOperations): Promise<AvmEphemeralForest> {
+    const treeMap = new Map<MerkleTreeId, EphemeralAvmTree>();
+    const treeTypes = [MerkleTreeId.NULLIFIER_TREE, MerkleTreeId.NOTE_HASH_TREE, MerkleTreeId.PUBLIC_DATA_TREE];
+    for (const treeType of treeTypes) {
+      const treeInfo = await treeDb.getTreeInfo(treeType);
+      const tree = await EphemeralAvmTree.create(treeInfo.size, treeInfo.depth, treeDb, treeType);
+      treeMap.set(treeType, tree);
+    }
+    return new AvmEphemeralForest(treeDb, treeMap, new Map(), new Map());
+  }
+
+  fork(): AvmEphemeralForest {
+    return new AvmEphemeralForest(
+      this.treeDb,
+      cloneDeep(this.treeMap),
+      cloneDeep(this.indexedTreeMin),
+      cloneDeep(this.indexedUpdates),
+    );
+  }
+
+  /**
+   * Gets sibling path for a leaf - if the sibling path is not found in the tree, it is fetched from the DB
+   * @param treeId - The tree to be queried for a sibling path.
+   * @param index - The index of the leaf for which a sibling path should be returned.
+   * @returns The sibling path of the leaf.
+   */
+  async getSiblingPath(treeId: MerkleTreeId, index: bigint): Promise<Fr[]> {
+    const tree = this.treeMap.get(treeId)!;
+    let path = tree.getSiblingPath(index);
+    if (path === undefined) {
+      // We dont have the sibling path in our tree - we have to get it from the DB
+      path = (await this.treeDb.getSiblingPath(treeId, index)).toFields();
+      // Since the sibling path could be outdated, we compare it with nodes in our tree
+      // if we encounter a mismatch, we replace it with the node we found in our tree.
+      for (let i = 0; i < path.length; i++) {
+        const siblingIndex = index ^ 1n;
+        try {
+          const node = tree.getNode(siblingIndex, tree.depth - i);
+          const nodeHash = tree.hashTree(node, i + 1);
+          if (!nodeHash.equals(path[i])) {
+            path[i] = nodeHash;
+          }
+        } catch (e) {
+          // Nothing happens here, we just didnt find the node in our tree
+          // So we dont change the value at this path index
+        }
+        index >>= 1n;
+      }
+    }
+    return path;
+  }
+
+  /**
+   * This does the work of appending the new leaf and updating the low witness
+   * @param treeId - The tree to be queried for a sibling path.
+   * @param lowWitnessIndex - The index of the low leaf in the tree.
+   * @param lowWitness - The preimage of the low leaf.
+   * @param newLeaf - The preimage of the new leaf to be inserted.
+   * @returns The sibling path of the new leaf (i.e. the insertion path)
+   */
+  async appendIndexedTree<ID extends IndexedTreeId, T extends IndexedTreeLeafPreimage>(
+    treeId: ID,
+    lowWitnessIndex: bigint,
+    lowWitness: T,
+    newLeaf: T,
```

Comment:
These are both preimages. Consider calling them `lowLeafPreimage` and `newLeafPreimage`. I think we use the term "leaf" to refer to the hash of the preimage.

Reply:
Same with lowWitnessIndex, maybe change to lowLeafIndex?